### PR TITLE
add bloom filter index

### PR DESCRIPTION
**Pull request by kangpinghuang**
Add bloom filter index to prune data when loading segment

### CODE REVIEW

File: be/src/olap/types.h [hunk 1]
```diff
@@ -74,6 +74,7 @@ class TypeInfo {
     inline void set_to_min(void* buf) const { _set_to_min(buf); }
 
     inline uint32_t hash_code(const void* data, uint32_t seed) const { return _hash_code(data, seed); }
+    inline uint64_t hash_code64(const void* data, uint64_t seed) const { return _hash_code64(data, seed); }
```
Comment (imay) [inline be/src/olap/types.h]: why do we need hash64?

File: be/src/olap/bloom_filter.hpp [hunk 2]
```diff
@@ -71,6 +70,16 @@ class BitSet {
         return true;
     }
 
+    bool init_with_deep_copy(uint64_t* data, uint32_t data_len) {
+        _data = new(std::nothrow) uint64_t[data_len];
```
Comment (imay) [inline be/src/olap/bloom_filter.hpp]: Who will delete allocated _data?

File: be/src/olap/rowset/segment_v2/bloom_filter_page.cpp [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+Status BloomFilterPageBuilder::add(const uint8_t* vals, size_t count) {
+    for (int i = 0; i < count; ++i) {
+        if (vals == nullptr) {
+            _bf_builder.add_bytes(nullptr, 1);
+        } else {
+            uint64_t hash = _type_info->hash_code64(vals, DEFAULT_SEED);
+            _bf_builder.add_hash(hash);
+        }
+        
+        if (vals != nullptr) {
```
Comment (imay) [inline be/src/olap/rowset/segment_v2/bloom_filter_page.cpp]: better to put this in else block above

File: be/src/olap/rowset/segment_v2/bloom_filter_page.cpp [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    Slice data((char*)_bf_builder.bit_set_data(), _bf_builder.bit_set_data_len() * 8);
+    size_t num = 1;
+    RETURN_IF_ERROR(_page_builder->add((const uint8_t*)&data, &num));
+    // reset the bloom filter builder
+    _bf_builder.clear();
+    _num_inserted = 0;
+    return Status::OK();
+}
+
+Slice BloomFilterPageBuilder::finish() {
```
Comment (imay) [inline be/src/olap/rowset/segment_v2/bloom_filter_page.cpp]:
```suggestion
Status BloomFilterPageBuilder::finish(Slice*) {
```

File: be/src/olap/rowset/segment_v2/bloom_filter_page.cpp [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+            flush();
+        }
+    }
+    return Status::OK();
+}
+
+Status BloomFilterPageBuilder::flush() {
+    Slice data((char*)_bf_builder.bit_set_data(), _bf_builder.bit_set_data_len() * 8);
+    size_t num = 1;
+    RETURN_IF_ERROR(_page_builder->add((const uint8_t*)&data, &num));
```
Comment (imay) [inline be/src/olap/rowset/segment_v2/bloom_filter_page.cpp]: How about when _page_builder is full?

Comment (kangpinghuang) [reply be/src/olap/rowset/segment_v2/bloom_filter_page.cpp]: here, I set config's data_page_size to 0 of page_builder, it will no be full.

**Review by gaodayue — CHANGES_REQUESTED** [review]
Besides implementation, I have several questions about the design

1. Why do we choose to store one BF per column per row block (1024 rows)? This seems to have a fairly high storage and search overhead. For example, let's say a column contains 10M rows and thus 10K blocks. Under the default fpp which 5%, the size of each BF is around  800 bytes. So the total size of this BF index is ~8MB which is very large(for comparison, the uncompressed size of int column with 10M rows is 40MB). To evaluate a "in" predicate with `n` values, `10K` BFs need to be tested on `n` values, which could also take a long time.
2. Size of BF depends on `expected_entries`, which is the number of unique values that will get inserted. Currently `expected_entries` is set to num_rows_per_block, which is too large for row block that contains many repeated values. How about allocating one BF for k distinct values instead of k rows? We can use the BF we're building to de-duplicate inputs.
3. Classic BF (the one Doris is currently using) is known to be cache unfriendly. Have you considered BlockedBloomFilter which is used by Parquet and RocksDB? For more information on BlockedBloomFilter, you can refer to https://github.com/apache/parquet-format/blob/master/BloomFilter.md

File: be/src/olap/types.h (hunk 1 above)
Comment (kangpinghuang) [reply be/src/olap/types.h]: it used to calculate hash code of specific type data, used by bloom filter.

File: be/src/olap/bloom_filter.hpp (hunk 2 above)
Comment (kangpinghuang) [reply be/src/olap/bloom_filter.hpp]: Bitset will free the allocated _data in destructor

**Comment by kangpinghuang** [comment]
> Besides implementation, I have several questions about the design
> 
> 1. Why do we choose to store one BF per column per row block (1024 rows)? This seems to have a fairly high storage and search overhead. For example, let's say a column contains 10M rows and thus 10K blocks. Under the default fpp which 5%, the size of each BF is around  800 bytes. So the total size of this BF index is ~8MB which is very large(for comparison, the uncompressed size of int column with 10M rows is 40MB). To evaluate a "in" predicate with `n` values, `10K` BFs need to be tested on `n` values, which could also take a long time.
> 2. Size of BF depends on `expected_entries`, which is the number of unique values that will get inserted. Currently `expected_entries` is set to num_rows_per_block, which is too large for row block that contains many repeated values. How about allocating one BF for k distinct values instead of k rows? We can use the BF we're building to de-duplicate inputs.
> 3. Classic BF (the one Doris is currently using) is known to be cache unfriendly. Have you considered BlockedBloomFilter which is used by Parquet and RocksDB? For more information on BlockedBloomFilter, you can refer to https://github.com/apache/parquet-format/blob/master/BloomFilter.md

good suggestion! Tks

File: be/src/olap/rowset/segment_v2/bloom_filter_page.cpp (hunk 4 above)
Comment (kangpinghuang) [reply be/src/olap/rowset/segment_v2/bloom_filter_page.cpp]: ok

File: be/src/olap/rowset/segment_v2/binary_plain_page.h [hunk 6]
```diff
@@ -152,7 +153,12 @@ class BinaryPlainPageDecoder : public PageDecoder {
 
         // Decode trailer
         _num_elems = decode_fixed32_le((const uint8_t *)&_data[_data.get_size() - sizeof(uint32_t)]);
-        _offsets_pos = _data.get_size() - (_num_elems + 1) * sizeof(uint32_t);
+        uint32_t tail_length = (_num_elems + 1) * sizeof(uint32_t);
+        if (_data.get_size() <= tail_length) {
```
[the remark this answers is not in the document]
Comment (kangpinghuang) [reply be/src/olap/rowset/segment_v2/binary_plain_page.h]: ok

File: be/src/olap/rowset/segment_v2/column_writer.h [hunk 7]
```diff
@@ -42,7 +44,13 @@ struct ColumnWriterOptions {
     // store compressed page only when space saving is above the threshold.
     // space saving = 1 - compressed_size / uncompressed_size
     double compression_min_space_saving = 0.1;
+    // for zone map index
     bool need_zone_map = false;
+
+    // for bloom filter index
+    bool is_bf_column = false;
+    size_t bloom_filter_block_size = 1024;
+    double fpp = BLOOM_FILTER_DEFAULT_FPP;
```
[the remark this answers is not in the document]
Comment (kangpinghuang) [reply be/src/olap/rowset/segment_v2/column_writer.h]: ok

File: be/src/olap/rowset/segment_v2/column_writer.h [hunk 8]
```diff
@@ -42,7 +44,13 @@ struct ColumnWriterOptions {
     // store compressed page only when space saving is above the threshold.
     // space saving = 1 - compressed_size / uncompressed_size
     double compression_min_space_saving = 0.1;
+    // for zone map index
     bool need_zone_map = false;
+
+    // for bloom filter index
+    bool is_bf_column = false;
```
[the remark this answers is not in the document]
Comment (kangpinghuang) [reply be/src/olap/rowset/segment_v2/column_writer.h]: ok

File: be/src/olap/rowset/segment_v2/segment_writer.cpp [hunk 9]
```diff
@@ -137,6 +139,13 @@ Status SegmentWriter::_write_zone_map() {
     return Status::OK();
 }
 
+Status SegmentWriter::_write_bloom_filter() {
+    for (auto column_writer : _column_writers) {
```
[the remark this answers is not in the document]
Comment (kangpinghuang) [reply be/src/olap/rowset/segment_v2/segment_writer.cpp]: ok

**Pull request closed without merging**